pkg/models/retry.go

```diff
@@ -132,23 +186,83 @@ func WithOnRetry(onRetry OnRetryFunc) RetryOption {
 	}
 }
 
+// WithOnRetry sets the callback function to execute on each retry.
+func WithOnRetryV2(onRetry OnRetryV2Func) RetryOption {
+	return func(cfg *RetryConfig) {
+		cfg.onRetryV2 = onRetry
+	}
+}
+
 // WithRetryIf sets the condition to determine whether to retry based on the error.
 func WithRetryIf(retryIf RetryIfFunc) RetryOption {
 	return func(cfg *RetryConfig) {
 		cfg.retryIf = retryIf
 	}
 }
 
+// WithRetryIfV2 sets the condition to determine whether to retry based on the response.
+// This enables the new response-based retry logic and automatically enables the
+// new behavior flags.
+func WithRetryIfV2(retryIf RetryIfFuncV2) RetryOption {
+	return func(cfg *RetryConfig) {
+		cfg.retryIf = func(err error) bool { return false } // Disable legacy retryIf
+		cfg.onRetry = func(n uint, err error) {}            // Disable legacy onRetry
+		cfg.retryIfV2 = retryIf
+		cfg.returnHTTPStatusAsErr = false // Enable new behavior by default when using V2
+	}
+}
+
+// WithReturnHTTPStatusAsErr controls the legacy behavior where HTTP status codes
+// are converted to Go errors. When enabled (true, default), HTTP status codes
+// are converted to errors (legacy behavior). When disabled (false), only actual
+// network/request errors are returned as errors, while HTTP responses (even with
+// 4xx/5xx status) return nil error (new correct behavior).
+func WithReturnHTTPStatusAsErr(enabled bool) RetryOption {
+	return func(cfg *RetryConfig) {
+		cfg.returnHTTPStatusAsErr = enabled
+	}
+}
+
 // Custom wrapper for http.Client that implements the Doer interface.
 // - Do
 // - DoWithRetry
 type HttpClient struct {
-	httpClient *http.Client
+	httpClient  *http.Client
+	retryConfig *RetryConfig
```

Comment:
Note: we add the RetryConfig to the struct as a way to influence the Client retry logic from the outside.